### PR TITLE
Vault enhancements

### LINKED ISSUE #7: Vault script - Add in a check to see if the 'server' keys have been CA Signed
Are we able to check if the 'server' keys have been updated from being self-signed to being CA signed? With the keys being signed by the CA as I understand it the integrity of the keys are improved, by default SHA1 is used however if they are signed by the CA then this can be improved SHA 2-xxx.

If self signed keys are used then the following warning appears in the italog after each service restart:

ITATP151W Security warning - The Signature Hash Algorithm of the Vault certificate is SHA1. We recommend that you use at least Signature Hash Algorithm SHA2-256 with this Vault's configuration.

### LINKED ISSUE #8: Vault script - issue with 'Vault_StaticIP' function - windows 2016 - vault v12
When running the script on the vault an error is seen when running the function 'Vault_StaticIP'. The function uses the following:

`Get-NetIPAddress -InterfaceAlias "Ethernet" -AddressFamily IPv4`

With the vault in my lab the 'InterfaceAlias' name is 'Ethernet0', I suspect that this will vary between vaults for different customers.

Perhaps missing out the '-InterfaceAlias' will help to shorten the output, this output will obviously be different as well when customers have multiple network cards and are in a vault cluster?

Perhaps list all the network interfaces in the output that have an IPv4 address (minus the loopback interface) and state which adapters do not have statically assigned IP's?

### LINKED ISSUE #9: Vault script - Check no additional firewall rules exist in the windows firewall
Would be good to check the firewall to see if there have been any manually added firewall rules added to the outgoing or incoming ranges.

Not sure how complicated the script needs to be for this check, I used the following before for a healthcheck to get the list or rules to a csv:

`function GetFirewallRules
{
$Direction = "*"
$RuleState = "*"
$Action = "*"
$hostname= hostname
$CSVFile = "$vComputerName-FirewallRules.csv"

$FirewallRules = Get-NetFirewallRule -DisplayName * -PolicyStore "ActiveStore"

$FirewallRuleSet = @()
ForEach ($Rule In $FirewallRules)
{
	Write-Output "Processing rule `"$($Rule.DisplayName)`" ($($Rule.Name))"

	$AdressFilter = $Rule | Get-NetFirewallAddressFilter
	$PortFilter = $Rule | Get-NetFirewallPortFilter
	$ApplicationFilter = $Rule | Get-NetFirewallApplicationFilter
	$ServiceFilter = $Rule | Get-NetFirewallServiceFilter
	$InterfaceFilter = $Rule | Get-NetFirewallInterfaceFilter
	$InterfaceTypeFilter = $Rule | Get-NetFirewallInterfaceTypeFilter
	$SecurityFilter = $Rule | Get-NetFirewallSecurityFilter

	$HashProps = [PSCustomObject]@{
		Name = $Rule.Name
		DisplayName = $Rule.DisplayName
		Description = $Rule.Description
		Group = $Rule.Group
		Enabled = $Rule.Enabled
		Profile = $Rule.Profile
		Platform = StringArrayToList $Rule.Platform
		Direction = $Rule.Direction
		Action = $Rule.Action
		EdgeTraversalPolicy = $Rule.EdgeTraversalPolicy
		LooseSourceMapping = $Rule.LooseSourceMapping
		LocalOnlyMapping = $Rule.LocalOnlyMapping
		Owner = $Rule.Owner
		LocalAddress = StringArrayToList $AdressFilter.LocalAddress
		RemoteAddress = StringArrayToList $AdressFilter.RemoteAddress
		Protocol = $PortFilter.Protocol
		LocalPort = StringArrayToList $PortFilter.LocalPort
		RemotePort = StringArrayToList $PortFilter.RemotePort
		IcmpType = StringArrayToList $PortFilter.IcmpType
		DynamicTarget = $PortFilter.DynamicTarget
		Program = $ApplicationFilter.Program -Replace "$($ENV:SystemRoot.Replace("\","\\"))\\", "%SystemRoot%\" -Replace "$(${ENV:ProgramFiles(x86)}.Replace("\","\\").Replace("(","\(").Replace(")","\)"))\\", "%ProgramFiles(x86)%\" -Replace "$($ENV:ProgramFiles.Replace("\","\\"))\\", "%ProgramFiles%\"
		Package = $ApplicationFilter.Package
		Service = $ServiceFilter.Service
		InterfaceAlias = StringArrayToList $InterfaceFilter.InterfaceAlias
		InterfaceType = $InterfaceTypeFilter.InterfaceType
		LocalUser = $SecurityFilter.LocalUser
		RemoteUser = $SecurityFilter.RemoteUser
		RemoteMachine = $SecurityFilter.RemoteMachine
		Authentication = $SecurityFilter.Authentication
		Encryption = $SecurityFilter.Encryption
		OverrideBlockRules = $SecurityFilter.OverrideBlockRules
	}

	$FirewallRuleSet += $HashProps
}

$FirewallRuleSet | ConvertTo-CSV -NoTypeInformation -Delimiter ";" | Set-Content "$Collect\System Information\$CSVFile"

}`

### PR DESCRIPTION
**Pull request by AssafMiron**
### What does this PR do?
- Adding more hardening checks for the Vault
- Fixing some issues

### What ticket does this PR close?
Resolves #7 
Resolves #9 
Might resolve #8

### Checklists
Run a test on a VM vault (to reproduce the Ethernet issue)
Run a test on a non hardened, default vault (to check the FW rules and Vault Certificate)

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs

### CODE REVIEW

**Comment by AssafMiron** [comment]
@itsmegb , @NathTheDude would be great if you can verify this new change as it should solve some of the enhancements and a bug that you identified for the Vault
Thanks!

**Comment by itsmegb** [comment]
This all looks good to me! Potentially an issue with the Firewall rule checking, the verbose logs report that i have 1 non-standard rule in my dbparm.ini but also states that i have 3 non-standard FW rules configured. Looking at the FW rules, i have 2 manual rules defined.
Would it be a good idea to clarify this a bit further?

**Review by itsmegb — CHANGES_REQUESTED** [review]
This all looks good to me! Potentially an issue with the Firewall rule checking, the verbose logs report that i have 1 non-standard rule in my dbparm.ini but also states that i have 3 non-standard FW rules configured. Looking at the FW rules, i have 2 manual rules defined.
Would it be a good idea to clarify this a bit further?

**Comment by AssafMiron** [comment]
Thanks @itsmegb 
I added also the ICMP rules that were missing in the DBParm.ini (these are configured behind the scenes and not as part of the file
So I suspect this is where the confusion might lay
ley me know if now it makes more sense

Thanks,
Assaf

**Comment by AssafMiron** [comment]
@itsmegb I added another change to the Static IP check to handle the special case of NIC teaming
We actually check if there is teaming configured in the NIC hardening check, so I  used that to check the Team name when checking for static IP configuration
If you still have your environment with NIC teaming configured, let me know if that works

Thanks,
Assaf

**Comment by AssafMiron** [comment]
@itsmegb after our talk (and a few debugging tries) I figured out the issue and updated the script
I checked it on my lab - please verify at yours

**Review by itsmegb — CHANGES_REQUESTED** [review]
There appears to be a line that breaks the script, in the new Vault_KeysProtection function.

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 [hunk 1]
```diff
@@ -331,4 +338,352 @@ Function Vault_LogicContainerServiceLocalUser
...
+		$tmpStatus = ""
+	}
+	Process {
+		try{
+			Write-LogMessage -Type Info -Msg "Start validating Vault Encryption keys permissions"
+			$vaultFolder = $(Get-DetectedComponents -Component Vault).Path
+			$DBParmFile = $(Get-ChildItem -Path $vaultFolder -Include "DBParm.ini" -Recurse).FullName
+			# Get the location of all Vault Keys
+			$keysList = $(Get-Content -Path $DBParmFile | Select-String -List "RecoveryPubKey","ServerKey","ServerPrivateKey","RecoveryPrvKey","BackupKey").Line
+			Write-LogMessage -Type Verbose -Msg "Found the following Keys paths: $($KeysList -join `n)"
```
[the remark this answers is not in the document]
Comment (AssafMiron) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:628]: Sorry about that, missed it somehow
Will push a new commit to fix it

[the remark this answers is not in the document]
Comment (itsmegb) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1]: Excellent, that is now working as expected.

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 [hunk 2]
```diff
@@ -625,9 +625,19 @@ Function Vault_KeysProtection
...
 			# Get the location of all Vault Keys
 			$keysList = $(Get-Content -Path $DBParmFile | Select-String -List "RecoveryPubKey","ServerKey","ServerPrivateKey","RecoveryPrvKey","BackupKey").Line
-			Write-LogMessage -Type Verbose -Msg "Found the following Keys paths: $KeysList"
+			Write-LogMessage -Type Verbose -Msg "Found the following Keys paths: $($KeysList -join `n)"
 			$KeysLocations = @()
 			$KeysLocations += $($keysList | % { Split-Path -Parent -Path $($_.Split("=")[1]) } ) | Select-Object -Unique
+			
+			# Check if the Recovery key exists on the server
+			$RecoveryKey = ($keysList | Where-Object { $_ -match "RecoveryPrvKey=" })
+			If(Test-Path $RecoveryKey)
```
[the remark this answers is not in the document]
Comment (AssafMiron) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:634]: I added more logging to verify
On my env I don't have it on the machine and I can see it in the log

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 [hunk 3]
```diff
@@ -588,3 +588,92 @@ Function Vault_ServerCertificate
...
+				{
+					$KeysFolderLocalSystem = $false
+					$res = "Warning"
+				}
+				$tmpStatus += "<li>" + $myRef.Value + "</li>"
+
+				# Verify if Administrators, System and the CPM User are the only ones that has permissions
+				if(($KeysFolderLocalAdmins -eq $true) -and ($KeysFolderLocalSystem -eq $true))
+				{
+					If((Compare-AmountOfUserPermissions -Path $path -amount 2 -outStatus ([ref]$myRef)) -ne "Good")
```
[the remark this answers is not in the document]
Comment (AssafMiron) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:661]: I think that Creator owner is not really a valid user that we would want
Maybe we should push customers to remove it
WDYT?

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (itsmegb) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:636]: So it looks like the value of $RecoveryKey includes the "RecoveryPrvKey=" part, so the test is failing as it isn't a valid location on disk.

I think just splitting the value returned by the match _should_ make the test work correctly.

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (itsmegb) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:666]: Removing CREATOR OWNER doesn't seem to effect the Vault (in my _very_ limited testing).

It does appear to be an inherited permission, so i'm not sure whether it is required or not?

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (AssafMiron) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:636]: Ahhh... you are correct and I can see that it was missing.
Adding now the split to have the real path

File: CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (AssafMiron) [reply CYBRHardeningCheck/Vault/VaultHardeningSteps.psm1:666]: Looking online a bit I see a lot of places saying to remove the Creator Owner permission
So I think it should be OK

**Pull request merged**